### PR TITLE
add_qVoxFall_plugin

### PR DESCRIPTION
**Pull request by farmakis**
We would like to contribute a plugin created by our research group at  the univrsity of Newcastle, NSW, Australia as part of an open pit mine monitoring program.
It concerns rockfall detection and volume estimation between two 3D models using a voxel grid.

### CODE REVIEW

File: plugins/private/qVoxFall/include/qVoxFallDialog.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+//#  This program is free software; you can redistribute it and/or modify  #
+//#  it under the terms of the GNU General Public License as published by  #
+//#  the Free Software Foundation; version 2 of the License.               #
+//#                                                                        #
+//#  This program is distributed in the hope that it will be useful,       #
+//#  but WITHOUT ANY WARRANTY; without even the implied warranty of        #
+//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
+//#  GNU General Public License for more details.                          #
+//#                                                                        #
+//#                             COPYRIGHT: XXX                             #
```
Comment (dgirardeau) [inline plugins/private/qVoxFall/include/qVoxFallDialog.h:14]: Missing copyright

File: plugins/private/qVoxFall/include/qVoxFall.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
+//#  GNU General Public License for more details.                          #
+//#                                                                        #
+//#                 COPYRIGHT: THE UNIVERSITY OF NEWCASTLE                 #
+//#                                                                        #
+//##########################################################################
+
+#pragma once
+
+#ifndef Q_VOXFALL_PLUGIN_HEADER
```
Comment (dgirardeau) [inline plugins/private/qVoxFall/include/qVoxFall.h:20]: Wherever there's a _#pragma once_ statement, you don't need the _#ifnded Q_XXX + #define Q_XXX + #endif_ statements (I think it applies to pretty much all header files?)

File: plugins/private/qVoxFall/src/qVoxFall.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+			mesh1->setEnabled(false);
+			m_app->dispToConsole("[VoxFall] Completed!", ccMainAppInterface::STD_CONSOLE_MESSAGE);
+		}	
+	}
+
+	//'Compute' may change some parameters of the dialog
+	dlg.saveParamsToPersistentSettings();
+}
+
+
```
Comment (dgirardeau) [inline plugins/private/qVoxFall/src/qVoxFall.cpp:123]: a lot of spurious blank lines?

File: plugins/private/qVoxFall/src/qVoxFallDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+void qVoxFallDialog::saveParamsTo(QSettings& settings)
+{
+	//save parameters
+	settings.setValue("VoxelSize", voxelSizeDoubleSpinBox->value());
+	settings.setValue("Azimuth", azDoubleSpinBox->value());
+	settings.setValue("ExportMeshesEnabled", exportCheckBox->isChecked());
+	settings.setValue("LossGainEnabled", lossCheckBox->isChecked());
+}
+
+
```
Comment (dgirardeau) [inline plugins/private/qVoxFall/src/qVoxFallDialog.cpp:217]: spurious blank lines

**Review by dgirardeau — CHANGES_REQUESTED** [review]
Thanks very much for the contribution!

Apart for a few syntax comments, my main comment would be that we prefer now to keep the code of these plugins in their original repository (so that you keep the ownership of the code), and we simply point to it as a submodule. Can you reorganize the files so that the root of the repository is the qVoxFall folder? You can create a new repository to not overwrite everything in the existing one.

See how it's done with the MPlane plugin for instance: https://github.com/hvs-ait/mplane-plugin/ (we then add the repository as a submodule in plugins/core

**Comment by dgirardeau** [comment]
Also, do you have any documentation we can point to in the wiki? Or do you want to create the wiki page directly maybe? (in which case I can create an account for you to edit it)

**Comment by farmakis** [comment]
Thank you for the review.

We have created the repository as requested, including the suggested changes in the files.

As for the documentation, there is a relevant research article that has just gone through the review phase. Soon to be published.

**Comment by dgirardeau** [comment]
> Thank you for the review.
> 
> We have created the repository as requested, including the suggested changes in the files.
> 
> As for the documentation, there is a relevant research article that has just gone through the review phase. Soon to be published.

Thanks. What's the address of this new repository?

**Comment by farmakis** [comment]
> > Thank you for the review.
> > We have created the repository as requested, including the suggested changes in the files.
> > As for the documentation, there is a relevant research article that has just gone through the review phase. Soon to be published.
> 
> Thanks. What's the address of this new repository?

Here is the link to the repository:
https://github.com/farmakis/qVoxFall

**Comment by dgirardeau** [comment]
Thanks. Would you agree to give me some access rights to this repo so that I can push minor fixes? (and later it might come handy to do some maintenance work).

**Comment by farmakis** [comment]
> Thanks. Would you agree to give me some access rights to this repo so that I can push minor fixes? (and later it might come handy to do some maintenance work).

Absolutely. Thanks

**Comment by dgirardeau** [comment]
Thanks. Here's the PR: https://github.com/farmakis/qVoxFall/pull/1

**Comment by dgirardeau** [comment]
Here's the actual integration PR (as a submodule): https://github.com/CloudCompare/CloudCompare/pull/2061

**Comment by dgirardeau** [comment]
I'll close this one, but it's still interesting to look at the compilation issues on Linux and macOS

**Comment by farmakis** [comment]
> I'll close this one, but it's still interesting to look at the compilation issues on Linux and macOS

Yes, I do not really see what causes those compilation issues. We 'll have a look again.
We would also need to update the disclaimer text when the relevant article is released.

**Comment by farmakis** [comment]
Dear CC team,

Regarding the documentation for the qVoxFall plugin.
The paper is about to be published the following days. Could you please give me access to edit a corresponding wiki page where I can also point to the publication?

Thank you,

---------------------------------------------------
Dr. Ioannis Farmakis, Ph.D.
Postdoctoral Researcher
Sapienza Università di Roma
Email: ***@***.***

**Comment by dgirardeau** [comment]
@farmakis  Can you send a mail to admin@cloudcompare.org about that so that I can send you a wiki account?

**Pull request closed without merging**